pyomo/solvers/plugins/solvers/GAMS.py

```diff
@@ -0,0 +1,792 @@
+
+        # If False, GAMSSolver will only create and keep GAMS model file.
+        solve = opts.pop("solve", True)
+
+        # If None, GAMS will use default solver for model type.
+        solver = opts.pop("solver", None)
+
+        # If None, will chose from lp, nlp, mip, and minlp.
+        mtype = opts.pop("mtype", None)
+
+        # After solving, store level and marginal
+        # values in original model's variables
+        load_model = opts.pop("load_model", True)
+
+        # Print the results to sys.stdout
+        print_result = opts.pop("print_result", False)
+
+        # Keep tmp files: gms, lst, gdx, pf
+        keep_files = opts.pop("keep_files", False)
+
+        # Turn on the GAMS "holdfixed" model attribute
+        holdfixed = opts.pop("holdfixed", False)
+
+        if len(opts):
+            raise ValueError(
+                "GAMSSolver passed unrecognized solver options:\n\t" +
+                "\n\t".join("%s = %s" % (k,v) for k,v in iteritems(opts)))
+
+        if solver is not None:
+            if solver.upper() not in valid_solvers:
+                raise ValueError(
+                    "GAMSSolver passed unrecognized solver: %s" % solver)
+
+        if mtype is not None:
+            valid_mtypes = set([
+                'lp', 'qcp', 'nlp', 'dnlp', 'rmip', 'mip', 'rmiqcp', 'rminlp',
+                'miqcp', 'minlp', 'rmpec', 'mpec', 'mcp', 'cns', 'emp'])
+            if mtype.lower() not in valid_mtypes:
+                raise ValueError(
+                    "GAMSSolver passed unrecognized model type: %s" % mtype)
+            if (solver is not None and
+                mtype.upper() not in valid_solvers[solver.upper()]):
+                raise ValueError("GAMSSolver passed solver (%s) unsuitable for"
+                                 " given model type (%s)" % (solver, mtype))
+
+        # Flag to keep GAMS model file, implied to keep if output_filename given
+        keep_output = True
+        if output_filename is None:
+            output_filename = AlphaNumericTextLabeler()(model) + ".gms"
+            keep_output = keep_files
+
+        if symbolic_solver_labels and (labeler is not None):
+            raise ValueError("GAMSSolver: Using both the "
+                             "'symbolic_solver_labels' and 'labeler' "
+                             "options is forbidden")
+
+        if symbolic_solver_labels:
+            var_labeler = con_labeler = AlphaNumericTextLabeler()
+        elif labeler is None:
+            var_labeler = NumericLabeler('x')
+            con_labeler = NumericLabeler('c')
+        else:
+            var_labeler = con_labeler = labeler
+
+        var_list = []
+        symbolMap = SymbolMap()
+
+        def var_recorder(obj):
+            ans = var_labeler(obj)
+            var_list.append(ans)
+            return ans
+
+        def var_label(obj):
+            return symbolMap.getSymbol(obj, var_recorder)
+
+        with PauseGC() as pgc:

```

Comment:
The solver should *call* the GAMS writer and not block-copy the writer code here.